doc/developer/design/20230210_storage_persist_sink_redux.md

```diff
@@ -0,0 +1,161 @@
+           _.-´     |    .-´               |       \
+        .-´  .------+----|-------+---------|--------\-----.
+       /    /            |       |         |         \     \
+,--------------.   ,-----------------.     |     ,-----------------.
+| write_batches|   |  write_batches  |     |     |  write_batches  |
+| worker 0     |   | worker 1        |     |     | worker N        |
++-----+--------+   +-+---------------+     |     +--+--------------+
+       \              \                    |        /
+        `-.            `,                  |       /
+           `-._          `-.               |      /
+               `-._         `-.            |     /
+                   `---------. `-.         |    /
+                             +`---`---+-------------,
+                             | append_batches       |
+                             | one arbitrary worker |
+                             +------+---------------+
+```
+
+The different operators work like this:
+
+1. `mint_batch_descriptions` emits new batch descriptions whenever the frontier of source `Collection` advances.
+A batch description is a pair of `(previous_upper, upper)` that tells write operators which updates to write and
+in the end tells the append operator what frontiers to use for the `compare_and_append`. This is a single-worker operator.
+2. `write_batches` writes the source `Collection` to persist. It does this by creating a persist `BatchBuilder` for each
+timestamp, and writing every value that occurs at each timestamp to the corresponding `BatchBuilder`. This uses a bounded
+amount of memory. This does not yet append the batches to the persist shard, the update are only uploaded/prepared to be
+appended to a shard. We only emit finalized batches to the `append_batches` operator when that batch is contained within
+a batch descriptions that we learned about from `mint_batch_descriptions`.
+3. `append_batches` takes as input the minted batch descriptions and written batches. Whenever the frontiers sufficiently advance,
+we take a batch description and all the batches that belong to it and append it to the persist shard.
+This also updates the shared frontier information in `storage_state` to match the persist frontier.
+This way, the controller eventually learns of the upper frontier of data written to the persist shard.
+Note that failing to `compare_and_append` must be retried, as we current expect only 1 replica to be
+running this sink operation for each source `Collection`.
+
+### Notes on the implementation
+
+- It is needlessly expensive to create a `BatchBuilder` for each timestamp.
+  - A change to this implementation is to avoid spilling data to persist for timestamps
+  whose values use a small amount of memory. These can be consolidated into large batches in `append_batches`,
+  similar to how <https://github.com/MaterializeInc/materialize/pull/17526> works.
+  - We do this in the current implementation, so doing it from many workers, instead of 1,
```

Comment:
wait, doesn't this now write `O(num_workers*num_timestamps)` blobs to s3 where before it wrote `O(num_timestamps)`? that would be pretty meaningful

Reply:
Oh I didn't think about that, that is correct. The RQ tests in my draft pr seem to suggest it doesn't make a difference though (for 4 workers)

Reply:
it certainly could make a difference for how much things cost though! s3 PUTs are currently the dominant factor in our s3 spend

Reply:
@danhhz Because all the batches are using batch builders, won't the number of blobs still be ~roughly `the_full_size_of_the_timestamp/blob_size`, for large enough pieces of data?

I guess this means for collections with small pieces of data, we will have more puts, if the cluster is large than xsmall. We could certainly prioritize the optimization mentioned in this document where we in-memory buffer small pieces of data (I think the optimal choice would keep and in-memory buffer of `blob_size * number_of_workers`, but I think in practice we would have to choose `blob_size * some_constant`

Reply:
I referring to this optimization: https://github.com/MaterializeInc/materialize/pull/17526, which we can adjust for storage as well

Reply:
yeah, I think "for large enough pieces of data" is the issue there. you'd need to be around `num_workers * 128MiB` per timestamp for that to work out. that might be true for the initial snapshot of a postgres source, but it's unlikely to be true in the steady state for almost any source. tbh, if anything saves us here, it's probably the opposite, which is that persist has an optimization to not write any blobs when there are no updates

keep in mind that 17526 hasn't been merged yet, and in fact, Frank has shown some skepticism. so I dunno if there's anything that concretely needs to change here about the plan but we _certainly_ need to remove "is strictly better" from the text :)

I added some metrics to track our s3 get+put cost recently and a graph on the persist dash. it'd be straightforward to push your prototype impl to staging, run some experiments, and compare vs the baseline to see how much of an effect it would have and if we're willing to eat those costs in the short term

Reply:
> we certainly need to remove "is strictly better" from the text :)

This I agree with, and I'll add some color to the doc

> t'd be straightforward to push your prototype impl to staging, run some experiments, and compare vs the baseline to see how much of an effect it would have and if we're willing to eat those costs in the short term

I wonder what we consider a "canonical" experiment here; I certainly don't where to turn when I want a large throughput kafka source...that said, I'd be a relatively large loadgen source would do the trick...

Reply:
Good intuition @danhhz !!

For a medium, `TICK INTEVAL 10ms`, `SCALE FACTOR 10` auction loadgen source, Here is the before and after my multi-worker branch, on a medium replica (7 timely workers): https://g-2531aff0fb.grafana-workspace.us-east-1.amazonaws.com/d/m3U1U6ZVk/persist?orgId=1&var-datasource=cloud-staging%20us-east-1&var-version=All&var-env=environment-a81cd134-d747-42ea-8b75-44fbcb649a6d-0&var-pod=cluster-u2-replica-4-0&var-shard=All&from=now-3h&to=now

At least for this example, its not quite linear scaling, but its obviously more expensive. For sources like this, with >1 qps, and small rows, the "empty batches dont write to s3" optimization doesn't save us, as many rows will be consolidated into second chunks, then distributed (fairly evenly) among the workers.

What we can do is:
- **Note this tradeoff in the document** (this is the most important thing to do)
- I can test adding the in-memory optimization as discussed, and see if it improves the situation

Whats unclear to me still is:
- Do we feel its necessary to do a test like above with large records? My guess is that, no matter _what you choose as the configured boundary the in-memory optimization_, there is always going to exist a workload right on the edge that is going to use more s3 writes. This problem appears to me to be similar as "someone's rows are always a bit bigger than 1 part, so we "waste" writes, and doesn't seem to be broadly fixable unless we have the system dynamically tune itself based on the workload...

Does that make sense?

Reply:
> wait, shouldn't the experiment be on a source that trickles in a small but consistent amount of data? that would be the worst case for num_blobs overhead, no?

I could also try this, but I expect the "empty batches" optimization to save us here

Reply:
https://g-2531aff0fb.grafana-workspace.us-east-1.amazonaws.com/goto/112eWJJ4k?orgId=1

the `u2` pod is the "fast but small records" source and the `u7` pod is the "trickle with small records" source, and the middle section is when I deployed the multi-worker sink. You can see that the empty-batch optimization does appear to work for trickles of updates!